x-pack/filebeat/input/cometd/input_test.go

```diff
@@ -0,0 +1,279 @@
+		"auth.oauth2.client.id":     "DEMOCLIENTID",
+		"auth.oauth2.client.secret": "DEMOCLIENTSECRET",
+		"auth.oauth2.user":          "salesforce_user",
+		"auth.oauth2.password":      "pwd",
+		"auth.oauth2.token_url":     "https://example.com/token",
+	}
+	inputtest.AssertNotStartedInputCanBeDone(t, NewInput, &config)
+}
+
+func TestMakeEventFailure(t *testing.T) {
+	event := beat.Event{
+		Timestamp: time.Now().UTC(),
+		Fields: common.MapStr{
+			"event": common.MapStr{
+				"id":      "DEMOID",
+				"created": time.Now().UTC(),
+			},
+			"message": "DEMOBODYFAIL",
+		},
+		Private: "DEMOBODYFAIL",
+	}
+	assert.NotEqual(t, event, makeEvent("DEMOID", "DEMOBODY"))
+}
+
+func TestNewInput_Run(t *testing.T) {
+	eventsCh := make(chan beat.Event)
+
+	outlet := &mockedOutleter{
+		onEventHandler: func(event beat.Event) bool {
+			eventsCh <- event
+			return true
+		},
+	}
+	connector := &mockedConnector{
+		outlet: outlet,
+	}
+	var inputContext finput.Context
+
+	var msg bay.TriggerEvent
+	msg.Data.Event.ReplayID = 1234
+	msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+	msg.Channel = channelName
+
+	config := map[string]interface{}{
+		"channel_name":              channelName,
+		"auth.oauth2.client.id":     "client.id",
+		"auth.oauth2.client.secret": "client.secret",
+		"auth.oauth2.user":          "user",
+		"auth.oauth2.password":      "password",
+	}
+
+	r := http.HandlerFunc(oauth2Handler)
+	server := httptest.NewServer(r)
+	serverURL = server.URL
+	config["auth.oauth2.token_url"] = server.URL + "/token"
+
+	cfg := common.MustNewConfigFrom(config)
+
+	input, err := NewInput(cfg, connector, inputContext)
+	require.NoError(t, err)
+	require.NotNil(t, input)
+
+	input.Run()
+	for _, event := range []beat.Event{<-eventsCh} {
+		require.NoError(t, err)
+		assertEventMatches(t, msg, event)
+	}
+	server.Close()
+}
+
+func TestNewInput_Run_Wait(t *testing.T) {
+	eventsCh := make(chan beat.Event)
+
+	outlet := &mockedOutleter{
+		onEventHandler: func(event beat.Event) bool {
+			eventsCh <- event
+			return true
+		},
+	}
+	connector := &mockedConnector{
+		outlet: outlet,
+	}
+	var inputContext finput.Context
+
+	var msg bay.TriggerEvent
+	msg.Data.Event.ReplayID = 1234
+	msg.Data.Payload = []byte(`{"CountryIso": "IN"}`)
+	msg.Channel = channelName
+
+	config := map[string]interface{}{
+		"channel_name":              channelName,
+		"auth.oauth2.client.id":     "client.id",
+		"auth.oauth2.client.secret": "client.secret",
+		"auth.oauth2.user":          "user",
+		"auth.oauth2.password":      "password",
+	}
+
+	r := http.HandlerFunc(oauth2Handler)
+	server := httptest.NewServer(r)
+	serverURL = server.URL
+	config["auth.oauth2.token_url"] = server.URL + "/token"
+
+	cfg := common.MustNewConfigFrom(config)
+
+	input, err := NewInput(cfg, connector, inputContext)
+	require.NoError(t, err)
+	require.NotNil(t, input)
+
+	input.Run()
+
+	go func() {
+		time.Sleep(100 * time.Millisecond) // let input.Stop() be executed.
+		for range eventsCh {
```

Comment:
Nothing in this test actually checks that this for loop exited.

If you reverse which code is in the goroutine it will reveal several bugs:

```go
	go func() {
		time.Sleep(100 * time.Millisecond) // let input.Stop() be executed.
		input.Wait()
	}()

	for range eventsCh {
	}
```

1) It will find the "send on closed channel" bug I noticed elsewhere.
2) This test case will never exit.